### PR TITLE
Implement DynamicQueue

### PR DESCRIPTION
**Pull request by mihails-strasuns**
Intended for enhancing scheduler/epoll internals but I bet there will be other use cases too.

### CODE REVIEW

**Comment by mihails-strasuns** [comment]
TODO: ability to lock shrinking to ensure iteration + pushing can work

**Comment by mihails-strasuns** [comment]
Done, ready for review

File: src/ocean/util/container/queue/DynamicQueue.d [hunk 1]
```diff
@@ -0,0 +1,244 @@
+/*******************************************************************************
+
+    Wrapper for a common boilerplate to imitate queue using dynamic array.
+
+    Provides IQueue!(T) interfaces, stores items internally using dynamic
+    array of items. Will grow indefinitely as long as there is any spare
+    memory. Popped items will initially be marked as "free" and occasionally
+    whole array will be shifted to optimized used memory space.
+
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/queue/DynamicQueue.d:9]: Missing legal info.

File: src/ocean/util/container/queue/DynamicQueue.d [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+            pointer to the element pushed into the queue or null if the queue is
+            full.
+
+    ***************************************************************************/
+
+    override public T* push ( )
+    {
+        if (this.auto_shrink)
+        {
+            if (oldest > this.buffer.length / 3)
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/queue/DynamicQueue.d]: `this.oldest`

File: src/ocean/util/container/queue/DynamicQueue.d [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+        Returns:
+            total number of bytes used by queue (used space + free space)
+
+    ***************************************************************************/
+
+    override public ulong total_space ( )
+    {
+        return this.buffer.length * T.sizeof;
+    }
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/queue/DynamicQueue.d:165]: Do you think there'd be any reason to add a method to query the capacity of the buffer?

**Review by gavin-norman-sociomantic — COMMENTED** [review]
LGTM, apart from the few minor points.

File: src/ocean/util/container/queue/DynamicQueue.d (hunk 3 above)
Comment (mihails-strasuns) [reply src/ocean/util/container/queue/DynamicQueue.d:165]: `Buffer` struct currently does not expose it and in general it is beyond our control (runtime-defined). I can add it if you can think of at least one use case :)

Comment (gavin-norman-sociomantic) [reply src/ocean/util/container/queue/DynamicQueue.d:165]: Ah ok. No, not really.

**Review by gavin-norman-sociomantic — APPROVED** [review]

**Pull request merged**